{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Main Program Testing\n",
    "1. Inputs:\n",
    "    * XMI Structure Model: .../Input/XMI/[....xml]\n",
    "    * BOM: .../Input/BOM/Catalogs.xlsx\n",
    "2. Outputs:\n",
    "    * Excel: .../Output/InstanceTables/InstanceTable[date_time].xlsx\n",
    "    * Log: .../Output/Logs/Log[date_time].txt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "ename": "SyntaxError",
     "evalue": "invalid syntax (<ipython-input-1-5164477e1c10>, line 6)",
     "output_type": "error",
     "traceback": [
      "\u001b[1;36m  File \u001b[1;32m\"<ipython-input-1-5164477e1c10>\"\u001b[1;36m, line \u001b[1;32m6\u001b[0m\n\u001b[1;33m    <a rel=\"license\" href=\"http://creativecommons.org/licenses/by-sa/4.0/\"><img alt=\"Creative Commons License\" style=\"border-width:0\" src=\"https://i.creativecommons.org/l/by-sa/4.0/88x31.png\" /></a><br />This work is licensed under a <a rel=\"license\" href=\"http://creativecommons.org/licenses/by-sa/4.0/\">Creative Commons Attribution-ShareAlike 4.0 International License</a>.\u001b[0m\n\u001b[1;37m    ^\u001b[0m\n\u001b[1;31mSyntaxError\u001b[0m\u001b[1;31m:\u001b[0m invalid syntax\n"
     ]
    }
   ],
   "source": [
    "#Author: Habibi Husain Arifin\n",
    "#Created Date: 10 January 2019\n",
    "#Last Updated Date: 6 February 2020\n",
    "#Version: 1.0\n",
    "#<a rel=\"license\" href=\"http://creativecommons.org/licenses/by-sa/4.0/\"><img alt=\"Creative Commons License\" style=\"border-width:0\" src=\"https://i.creativecommons.org/l/by-sa/4.0/88x31.png\" /></a><br />This work is licensed under a <a rel=\"license\" href=\"http://creativecommons.org/licenses/by-sa/4.0/\">Creative Commons Attribution-ShareAlike 4.0 International License</a>.\n",
    "\n",
    "import datetime as DT\n",
    "import xml.etree.ElementTree as ET\n",
    "import copy\n",
    "import pandas as PD\n",
    "\n",
    "from deap import base, creator, tools, algorithms\n",
    "\n",
    "#Custom API\n",
    "from API.StructureModel import XMI25Element, Colors, XPathHelper, GeneticAlgorithmHelper\n",
    "from API.RNLTree import RNLTree\n",
    "from API.ExcelManager import ExcelManager\n",
    "from API.Instance import Instance\n",
    "from API.GeneticAlgorithms import GeneticAlgorithms\n",
    "from API.Configuration import Configuration\n",
    "\n",
    "#Change this one later\n",
    "MOE = \"MOE\"\n",
    "\n",
    "#File path\n",
    "SOI_NAME = \"Car\"\n",
    "#INPUT_XMI_FILE_PATH = \"Input/XMI/car-variants-150%-190-sp2-v1.0-hbb.xml\"\n",
    "#INPUT_XMI_FILE_PATH = \"Input/XMI/car-variants-100%-standard-190-sp2-v1.0-hbb.xml\"\n",
    "INPUT_XMI_FILE_PATH = \"Input/XMI/car-variants-100%-sport-190-sp4-v1.0-hbb.xml\"\n",
    "INPUT_CATALOG_FILE_PATH = \"Input/BOM/CarVariants-Catalogs.xlsx\"\n",
    "OUTPUT_EXCEL_FILE_PATH = \"Output/InstanceTables/CarVariants-InstanceTable\" + str(DT.datetime.now().strftime(\"%Y_%m_%d_%H_%M_%S\")) + \".xlsx\" \n",
    "OUTPUT_LOG_FILE_PATH = \"Output/Logs/CarVariants-Log\" + str(DT.datetime.now().strftime(\"%Y_%m_%d_%H_%M_%S\")) + \".txt\"  \n",
    "\n",
    "\"\"\"\n",
    "#Test Read StructureModel\n",
    "\"\"\"\n",
    "print(\"READ STRUCTURE MODEL\")\n",
    "tree = ET.parse(INPUT_XMI_FILE_PATH)\n",
    "root = tree.getroot()\n",
    "structModel = XMI25Element.createNewStructureModel(SOI_NAME, root)\n",
    "\n",
    "#Print to file\n",
    "#file = open(OUTPUT_FILE_PATH,\"w\")\n",
    "for obj in structModel:\n",
    "    print(Colors.getFontColorByLevel(obj.level), Colors.BOLD, XPathHelper.getRepeatedChar(\"\", obj.level), obj.stereotypes, \"LV.\", obj.level, Colors.ENDC, \", gene =\", obj.gaGene, \", nId =\", obj.nodeId, \", pId =\", obj.parentNodeId, \", xmiId =\", obj.xmiId, \", name =\", obj.name, \", value =\", obj.value, \", formula =\", obj.formula )\n",
    "    #file.write(\"\")\n",
    "print(\"\\nNo. of element:\", len(structModel), \"DateTime:\", str(DT.datetime.now()))\n",
    "\n",
    "#RNLTree.drawGraph(structModel)\n",
    "RNLTree.drawSimplifiedGraph(structModel, root)\n",
    "\n",
    "\"\"\"\n",
    "#Test Read Instance from ExcelFile\n",
    "\"\"\"\n",
    "print(\"READ INSTANCES FROM CATALOG/BOM FILE\")\n",
    "\n",
    "instances, maxRowSize, maxColSize = ExcelManager.readExcelFile(INPUT_CATALOG_FILE_PATH)\n",
    "instances = Instance.generateGeneSequentially(instances, maxRowSize)\n",
    "print(\"MaxRowSize:\", maxRowSize, \"MaxColSize:\", maxColSize)\n",
    "for instance in instances:\n",
    "    print(instance.insType, instance.gaGene, instance.parameters)\n",
    "    \n",
    "\"\"\"\n",
    "#Test Genetic Algorithm\n",
    "\"\"\"\n",
    "#GAParameters\n",
    "IND_SIZE = GeneticAlgorithmHelper.getIndividualSize(structModel) #We need to get it from the StructureModel\n",
    "SIZE_OF_GENE = Instance.getSizeOfGene(maxRowSize) #We need to get it from the BOM\n",
    "NO_OF_BEST_SOLUTIONS = 5 #Get it from the User UI\n",
    "NO_OF_GENERATIONS = 100 #Get it from the User UI\n",
    "POPULATION_SIZE = 10 #Get it from the User UI\n",
    "\n",
    "print(\"RUN GENETIC ALGORITHMS\")\n",
    "toolbox = base.Toolbox()\n",
    "\n",
    "creator.create(\"FitnessMax\", base.Fitness, weights=(1.0,))\n",
    "creator.create(\"Individual\", list, fitness=creator.FitnessMax)\n",
    "\n",
    "toolbox.register(\"attr_alphabet\", GeneticAlgorithms.generateGeneRandomly, k=SIZE_OF_GENE)\n",
    "toolbox.register(\"individual\", tools.initRepeat, creator.Individual, toolbox.attr_alphabet, n=IND_SIZE)\n",
    "toolbox.register(\"population\", tools.initRepeat, list, toolbox.individual)\n",
    "toolbox.register(\"evaluate\", GeneticAlgorithms.evaluation, structureModel=structModel, instances=instances)\n",
    "\n",
    "#Executes a one point crossover on the input sequence individuals.\n",
    "#The two individuals are modified in place.\n",
    "#The resulting individuals will respectively have the length of the other.\n",
    "toolbox.register(\"mate\", tools.cxOnePoint)\n",
    "\n",
    "#Shuffle the attributes of the input individual and return the mutant.\n",
    "#The individual is expected to be a sequence.\n",
    "#The indpb argument is the probability of each attribute to be moved.\n",
    "#Usually this mutation is applied on vector of indices.\n",
    "toolbox.register(\"mutate\", tools.mutShuffleIndexes, indpb=0.2)\n",
    "\n",
    "#Flip the value of the attributes of the input individual and return the mutant.\n",
    "#The individual is expected to be a sequence and the values of the attributes shall stay valid after the not operator is called on them.\n",
    "#The indpb argument is the probability of each attribute to be flipped.\n",
    "#This mutation is usually applied on boolean individuals.\n",
    "#toolbox.register(\"mutate\", tools.mutFlipBit, indpb=0.2)\n",
    "\n",
    "#This function applies a gaussian mutation of mean mu and standard deviation sigma on the input individual.\n",
    "#This mutation expects a sequence individual composed of real valued attributes.\n",
    "#The indpb argument is the probability of each attribute to be mutated.\n",
    "#toolbox.register(\"mutate\", tools.mutGaussian, mu=0, sigma=1, indpb=0.2)\n",
    "\n",
    "toolbox.register(\"select\", tools.selBest)\n",
    "\n",
    "#Tool decoration is a very powerful feature that helps to control very precise things during an evolution\n",
    "#without changing anything in the algorithm or operators.\n",
    "chromosomeSeq = XMI25Element.getChromosomeSequence(structModel)\n",
    "toolbox.decorate(\"mate\", GeneticAlgorithms.checkExistence(chromosomeSeq, instances))\n",
    "toolbox.decorate(\"mutate\", GeneticAlgorithms.checkExistence(chromosomeSeq, instances))\n",
    "\n",
    "pop = GeneticAlgorithms.customEvolution(POPULATION_SIZE, NO_OF_GENERATIONS, toolbox)\n",
    "bests = tools.selBest(pop, k=NO_OF_BEST_SOLUTIONS)\n",
    "\n",
    "#Print to file\n",
    "file = open(OUTPUT_LOG_FILE_PATH,\"w\")\n",
    "print(\"\\n############################################################\")\n",
    "print(\"\\n\", chromosomeSeq, \"\\nSolution\\tFitness\")\n",
    "file.write(\":\".join(chromosomeSeq) + \"\\nSolution\\tFitness\\n\")\n",
    "listChromosome = []\n",
    "listFitness = []\n",
    "listMOEs = []\n",
    "\n",
    "for best in bests:\n",
    "    print(\"\\n\", best, best.fitness)\n",
    "    file.write(\"\".join(best) + \"\\t\" + str(best.fitness) + \"\\n\")\n",
    "    listChromosome.append(\"\".join(best))\n",
    "    listFitness.append(str(best.fitness))\n",
    "    \n",
    "    #Create configuration for Solution(s)\n",
    "    sm = copy.deepcopy(structModel)\n",
    "    bestConf = Configuration.createNewConfiguration(sm, best, instances) #createNewConfiguration(structureModel, chromosome, instances)\n",
    "    moes = XMI25Element.getElementsByStereotype(bestConf.structureModel, MOE)\n",
    "    dictMOE = {}\n",
    "    for moe in moes:\n",
    "        print(moe.name, \":\", moe.value, moe.unitOnValueType)\n",
    "        file.write(str(moe.name) + \":\" + str(moe.value) + \" \" + str(moe.unitOnValueType) + \"\\n\")\n",
    "        dictMOE[str(moe.name) + \"::\" + str(moe.unitOnValueType)] = str(moe.value)\n",
    "    listMOEs.append(dictMOE)\n",
    "    dictMOE = {}\n",
    "file.close()\n",
    "\n",
    "#Write to excel\n",
    "dicts = {}\n",
    "dicts[\"Chromosome\" + \"[\" + \",\".join(chromosomeSeq) + \"]\"] = listChromosome\n",
    "dicts[\"Fitness Value\"] = listFitness\n",
    "for listMOE in listMOEs:\n",
    "    for key in listMOE.keys():\n",
    "        if key not in dicts:dicts[key] = []\n",
    "        dicts[key].append(listMOE[key])\n",
    "df = PD.DataFrame(dicts)\n",
    "ExcelManager.writeExcelFile(OUTPUT_EXCEL_FILE_PATH, df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
